src/language/extension-types.md

```diff
@@ -0,0 +1,315 @@
+You can choose to use some members, omit others,
+replace others, and add new functionality, 
+all while getting compile time errors before runtime errors.
+
+The following example wraps the `int` type to create an extension type
+that only allows operations that make sense for ID numbers:
+
+```dart
+extension type IdNumber(int i) {
+  // Wraps the 'int' type's '<' operator:
+  operator <(IdNumber other) => i < other.i;
+  // Doesn't declare the '+' operator, for example,
+  // because addition does not make sense for ID numbers.
+}
+
+void main() {
+  // Without the discipline of an extension type,
+  // 'int' exposes ID numbers to unsafe operations:
+  int myUnsafeId = 42424242;
+  myUnsafeId = myUnsafeId + 10; // OK, but shouldn't be allowed for IDs.
+
+  var safeId = IdNumber(42424242);
+  safeId + 10; // Compile-time error, no '+' operator.
+  myUnsafeId = safeId; // Compile-time error, wrong type.
+  myUnsafeId = safeId as int; // OK, runtime cast to representation type.
+}
+```
+
+{{site.alert.note}}
+Extension types serve the same purpose as **wrapper classes**,
+but don't require the creation of expensive runtime objects.
+Because extension types are static-only and compiled away at runtime,
+they are essentially zero cost.
+
+[**Extension methods**]() are a static abstraction similar to extension types.
+However, an extension method adds functionality to *all* instances
+of its underlying type. Extension types are the opposite; 
+an extension type's interface only applies to objects
+declared with the extension type,
+and is distinct from the interface of its underlying type by default. 
+Read the [Transparency](#transparency) section for more details.
+{{site.alert.end}}
+
+## Syntax
+
+### Declaration
+
+Define a new extension type with the `extension type` declaration and a name,
+followed by the *representation type declaration* in parenthesis:
+
+```dart
+extension type E(int i) {
+  // Define set of operations
+}
+```
+
+The representation type declaration specifies that 
+the underlying type of extension type `E` is `int`,
+and the reference to the *representation object* is named `i`.
+
+The representation object gives access to the underlying type,
+the same way declaring a final instance variable of the underlying type
+would in a wrapper class.
+It's in scope in the extension type body, and
+you can access it using its name as a getter:
+
+- Within the extension type body: `this.i`.
+- Outside with a property extraction: `e.i` (where `e` is the object of an extension type). 
+
+Extension types declarations can also include [type parameters](generics)
+just like classes:
+
+```dart
+extension type E<T>(int i) {
+  // ...
+}
+```
+
+### Constructors
+
+You can optionally declare constructors in an extension type's body.
+A constructor must be named, because the the representation declaration
+already acts as an unnamed constructor for the extension type
+that initializes the represenation object.
```

Comment:
@lrhn Sorry, I can't tell if this example is showing something different from the basic example in @eernstg's comment on the same thing. I mean there are obvious differences like the type parameters and making it `const`, but are those details relevant to naming the representation declaration, or did you just happen to include them because they were part of the actual extension type you wrote? Do you think we should include them here?